scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,45 @@
+
+- **Author(s):** Yi Li <yli3@redhat.com>
+- **Created:** [2023-10-10]
+
+## Status
+
+Accepted
+
+## Context
+
+The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 
+
+The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.
+
+
+## Decision
+
+The current NVD CVSS GitHub Workflow will be updated to fetch data from the NVD CVE API instead of NVD Json feeds while maintaining the freshness of the NVD CVSS data bundle in Google Storage. It will categorize the downloaded data by CVSS V3 severity levels, and these categorized data will be archived into individual zip files, such as `severity-low.zip`. The format of the JSON is the same as before.  
+
+We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.
+
+The CVSS Updater in Central will be modified to download zip files corresponding to four distinct severity levels and write them into a single zip file stored in Central's file system. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle and populate the Matcher DB with the enrichment data.
+
+The CVSS Updater will attempt to download each zip file from the Google bucket up to five times. If all five attempts fail, it will skip that particular file. For instance
+
+| Action                                                   | Result                                         | Next Step                               |
+|----------------------------------------------------------|------------------------------------------------|-----------------------------------------|
+| GET /bucket/1st-severity.zip                            | Success                                        | download to a tmp directory (disk)      |
+| GET /bucket/2nd-severity.zip                             | Success                                        | download to same tmp directory (disk)   |
+| GET /bucket/3rd-severity.zip                             | Failed (e.g. 401, 403, 505, Timeout)           | Retry 5 times and SKIP                  |
+| Create a final zip file from all jsons in downloaded zip files and save to store.Snapshot      | Success                                        | Wait 4 hours for another update         |
```

Comment:
we used to do 3 hours. We changed to 4 due to OpenShift CI's... efficiencies...

How often does Clair fetch vuln updates? Do we know about how often NVD's data may change (not sure if it's daily, hourly, weekly, etc)?

Reply:
https://nvd.nist.gov/vuln/data-feeds says:
The APIs are updated as frequently as our website (unlike the traditional feeds which have explicit update intervals)

https://nvd.nist.gov/general/FAQ-Sections/General-FAQs#faqLink8 also says:
Q. How often is the data in the NVD updated?
 A. CVEs are typically available in the NVD within an hour of their publishing. Once a CVE is in the NVD, analysts can begin the analysis process. The processing time can vary depending on the CVE, the information available, and the quantity of CVEs published within a given time frame. The NVD updates CVE records as soon as analysis is complete. CVE data and/or CPE names may later be updated . If modification occurs, the NVD will automatically refresh any relevant information.